crates/spidermonkey-embedding-splicer/src/stub_wasi.rs

```diff
@@ -77,16 +117,36 @@ pub fn stub_wasi(wasm: Vec<u8>, features: Vec<Features>) -> Result<Vec<u8>> {
 
     let has_io = features.contains(&Features::Clocks)
         || features.contains(&Features::Stdio)
-        || features.contains(&Features::Http);
+        || features.contains(&Features::Http)
+        || target_world_requires_io(&target_world_imports);
     if !has_io {
         stub_io(&mut module)?;
     }
 
-    stub_sockets(&mut module)?;
+    stub_sockets(&mut module, &target_world_imports)?;
 
     Ok(module.emit_wasm())
 }
 
+fn target_world_requires_io(target_world_imports: &HashSet<String>) -> bool {
+    return target_world_imports.contains("wasi:sockets/instance-network@0.2.0")
+        || target_world_imports.contains("wasi:sockets/udp@0.2.0")
+        || target_world_imports.contains("wasi:sockets/udp-create-socket@0.2.0")
+        || target_world_imports.contains("wasi:sockets/tcp@0.2.0")
+        || target_world_imports.contains("wasi:sockets/tcp-create-socket@0.2.0")
+        || target_world_imports.contains("wasi:sockets/ip-name-lookup@0.2.0")
+        || target_world_imports.contains("wasi:sockets/network@0.2.0")
+        || target_world_imports.contains("wasi:filesystem/types@0.2.0")
+        || target_world_imports.contains("wasi:filesystem/preopens@0.2.0")
+        || target_world_imports.contains("wasi:cli/environment@0.2.0")
+        || target_world_imports.contains("wasi:cli/exit@0.2.0")
+        || target_world_imports.contains("wasi:cli/terminal-stdin@0.2.0")
+        || target_world_imports.contains("wasi:cli/terminal-stdout@0.2.0")
+        || target_world_imports.contains("wasi:cli/terminal-stderr@0.2.0")
+        || target_world_imports.contains("wasi:cli/terminal-input@0.2.0")
+        || target_world_imports.contains("wasi:cli/terminal-output@0.2.0");
```

Comment:
`cli/environment` and `cli/exit` actually don't require IO:

```suggestion
    return target_world_imports.contains("wasi:sockets/instance-network@0.2.0")
        || target_world_imports.contains("wasi:sockets/udp@0.2.0")
        || target_world_imports.contains("wasi:sockets/udp-create-socket@0.2.0")
        || target_world_imports.contains("wasi:sockets/tcp@0.2.0")
        || target_world_imports.contains("wasi:sockets/tcp-create-socket@0.2.0")
        || target_world_imports.contains("wasi:sockets/ip-name-lookup@0.2.0")
        || target_world_imports.contains("wasi:sockets/network@0.2.0")
        || target_world_imports.contains("wasi:filesystem/types@0.2.0")
        || target_world_imports.contains("wasi:filesystem/preopens@0.2.0")
        || target_world_imports.contains("wasi:cli/terminal-stdin@0.2.0")
        || target_world_imports.contains("wasi:cli/terminal-stdout@0.2.0")
        || target_world_imports.contains("wasi:cli/terminal-stderr@0.2.0")
        || target_world_imports.contains("wasi:cli/terminal-input@0.2.0")
        || target_world_imports.contains("wasi:cli/terminal-output@0.2.0");
```